RavenDB-4323 Continue the work on PR #1567

fitzchak opened the pull request:
#1567

Commit:
RavenDB-4323 IndexDefinition.Compare should be aware of OutputReduceToCollection

Commit:
RavenDB-4323 Use also FromIndex in when flagging artificial document

Commit:
RavenDB-4323 Update schema version because of change to the tombstone schema

Commit:
RavenDB-4323 minor - fix naming

Commit:
RavenDB-4323 Extract output reduce to collection code out of IndexWriteOperation to a separate class.
Use LazyStringValue to hold the ReduceKeyHash.

Commit:
RavenDB-4323 Use the Lucene.FlushToDisk only when we do the commit for lucene and not also around the OutputReduceDocuments which has a standalone stats.

Commit:
RavenDB-4323 Fix NRE

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/OutputReduceIndexWriteOperation.cs:
```diff
@@ -0,0 +1,133 @@
+            using (stats.For(IndexingOperation.Reduce.SaveOutputDocuments))
+            {
+                var enqueue = DocumentDatabase.TxMerger.Enqueue(_outputReduceToCollectionCommand);
+                base.Commit(stats);
```
What happens if this throws?

fitzchak replied:
It will be handled as the higher level in indexing as before. 
Note that we had try..finally around this, not try..catch.

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/OutputReduceIndexWriteOperation.cs:
```diff
@@ -0,0 +1,133 @@
+            public class OutputReduceDocument
+            {
+                public bool IsDelete;
+                public LazyStringValue ReduceKeyHash;
```
Who owns the memory here?

fitzchak replied:
Working on this

Commit:
RavenDB-4323 Use a separate JsonOperationContext which would be long lived as long as needed by OutputReduceToCollectionCommand.
This also improve the previous code by not use json.net to convert to blitable but creating a DynamicJsonObject directly.
In regards to the document key we are using a string instead of a LazyStringValue because that we will anyway will use a string as context.ReadObject will take a string.

Commit:
RavenDB-4323 Do not nest the call for stats in order to be shown in the same level in the studio

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/OutputReduceIndexWriteOperation.cs:
```diff
@@ -54,80 +55,101 @@ public override void IndexDocument(LazyStringValue key, object document, Indexin
+                var key = _outputReduceToCollection + "/" + reduceKeyHash;
+
+                var djv = new DynamicJsonValue();
+                var propertyAccessor = PropertyAccessor.Create(reduceObject.GetType());
```
Shouldn't this be cached? See the impl in `AnonymousLuceneDocumentConverter.Getfields`

fitzchak replied:
https://github.com/ravendb/ravendb/pull/1606
